ouroboros-network/src/Ouroboros/Network/PeerSelection/PeerSelectionActions.hs

```diff
@@ -96,35 +115,56 @@ withPeerSelectionActions
   readPeerSharingController
   readNewInboundConnections
   peerStateActions
-  getLedgerPeers
+  ledgerPeersRng
+  ledgerPeersConsensusInterface
+  getUseLedgerAfter
   k = do
     localRootsVar <- newTVarIO mempty
+    reqVar  <- newEmptyTMVarIO
+    respVar <- newEmptyTMVarIO
+    let getRequest  = takeTMVar reqVar
+        putResponse = putTMVar  respVar
+        getLedgerPeers :: NumberOfPeers -> LedgerPeersKind -> m (Maybe (Set peeraddr, DiffTime))
+        getLedgerPeers = \numberOfPeers ledgerPeersKind -> do
+          atomically $ putTMVar reqVar (numberOfPeers, ledgerPeersKind)
+          atomically $ takeTMVar respVar
+
     let peerSelectionActions = PeerSelectionActions {
             readPeerSelectionTargets,
             readLocalRootPeers = readTVar localRootsVar,
             readNewInboundConnection = readNewInboundConnections,
             peerSharing,
             peerConnToPeerSharing,
-            requestPublicRootPeers,
-            requestBigLedgerPeers,
+            requestBigLedgerPeers = requestBigLedgerPeers getLedgerPeers,
+            requestPublicRootPeers = requestPublicRootPeers getLedgerPeers,
             requestPeerShare,
             peerStateActions
           }
+
     withAsync
-      (localRootPeersProvider
-        localRootTracer
-        toPeerAddr
-        DNS.defaultResolvConf
-        dnsActions
-        readLocalRootPeers
-        localRootsVar)
-      (\thread -> k thread peerSelectionActions)
+      (ledgerPeersThread ledgerPeersRng toPeerAddr ledgerPeersTracer getUseLedgerAfter
+                         ledgerPeersConsensusInterface dnsActions
+                         getRequest putResponse)
+      (\lpThread ->
```

Comment:
It's better to expose `with` style combinator rather than `ledgerPeersThread` directly, this allows to hide `getRequest` & `putResponse` as internal details.